Editor/EditorCore/PolyShapeTool.cs

```diff
@@ -0,0 +1,958 @@
+                    {
+                        if (SceneCameraIsAlignedWithPolyUp())
+                            SetPolyEditMode(PolyShape.PolyEditMode.Edit);
+                        else
+                            SetPolyEditMode(PolyShape.PolyEditMode.Height);
+
+                        evt.Use();
+                    }
+                    else if (polygon.polyEditMode == PolyShape.PolyEditMode.Height)
+                    {
+                        SetPolyEditMode(PolyShape.PolyEditMode.Edit);
+                        evt.Use();
+                    }
+                    else if (polygon.polyEditMode == PolyShape.PolyEditMode.Edit)
+                    {
+                        SetPolyEditMode(PolyShape.PolyEditMode.None);
+                        evt.Use();
+                    }
+
+                    break;
+                }
+
+                case KeyCode.Backspace:
+                {
+                    if (m_SelectedIndex > -1)
+                    {
+                        UndoUtility.RecordObject(polygon, "Delete Selected Points");
+                        polygon.m_Points.RemoveAt(m_SelectedIndex);
+                        m_SelectedIndex = -1;
+                        RebuildPolyShapeMesh(polygon);
+                        evt.Use();
+                    }
+                    break;
+                }
+
+                case KeyCode.Escape:
+                {
+                    polygon.m_Points.Clear();
+                    SetPolyEditMode(PolyShape.PolyEditMode.None);
+                    evt.Use();
+                    break;
+                }
+
+            }
+        }
+
+        /**
+         *  Is the scene camera looking directly at the up vector of the current polygon?
+         *  Prevents a situation where the height tool is rendered useless by coplanar
+         *  ray tracking.
+         */
+        bool SceneCameraIsAlignedWithPolyUp()
+        {
+            float dot = Vector3.Dot(SceneView.lastActiveSceneView.camera.transform.forward, polygon.transform.up);
+            return Mathf.Abs(Mathf.Abs(dot) - 1f) < .01f;
+        }
+
+        void OnSelectModeChanged(SelectMode selectMode)
+        {
+            // User changed select mode manually, remove InputTool flag
+            if (polygon != null
+                && polygon.polyEditMode != PolyShape.PolyEditMode.None
+                && !selectMode.ContainsFlag(SelectMode.InputTool))
+            {
+                SetPolyEditMode(PolyShape.PolyEditMode.None);
+            }
+        }
+
+        void OnBeginVertexMovement()
+        {
+            if (!m_IsModifyingVertices)
+                m_IsModifyingVertices = true;
+        }
+
+        void OnFinishVertexMovement()
+        {
+            m_IsModifyingVertices = false;
+            RebuildPolyShapeMesh(polygon);
+        }
+
+        void UndoRedoPerformed()
+        {
+            ClearLineRenderers();
+            InitLineRenderers();
```

Comment:
Now that this is an `EditorTool` is it necessary to `InitLineRenderers` on Undo still?

Reply:
I think it is as you can undo "add point to polyshape" thus the polyshape path must be updated accordingly.